JoodHajeer/assignment2/JoodHajeer-2.cpp

```diff
@@ -0,0 +1,252 @@
+#include <iostream>
+#include <stdio.h>  
+#include <string>
+#include <queue>
+#include <stack>
+#include <stdlib.h>
+using namespace std;
+
+template <class T>
+class Binary_Tree;
+template <class T>
+class BTNode {
+	//Binary tree Node Class building block for a BT
+	// A BT node class to store elements of a generic type T
+public:
+	BTNode(T &VAL, BTNode * LEFT = nullptr, BTNode * RIGHT = nullptr, BTNode * PARENT = nullptr, int DEPTH = -1);
+	~BTNode();
+private:
+	T val;
+	BTNode * left;
+	BTNode * right;
+	BTNode * parent;
+	int depth;
+	friend class Binary_Tree<T>;
+};
+template <class T>
+BTNode<T>::BTNode(T &VAL, BTNode * LEFT = nullptr, BTNode * RIGHT = nullptr, BTNode * PARENT = nullptr, int DEPTH = -1){
+	val = VAL;
+	left = LEFT;
+	right = RIGHT;
+	parent = PARENT;
+	depth = DEPTH;
+}
+template <class T>
+BTNode<T>::~BTNode() {}
+
+template <class T>
+class Binary_Tree{
+public:
+	Binary_Tree();
+	Binary_Tree( Binary_Tree <T> &copy_from);
+	~Binary_Tree();
+	bool insert(T &VAL);
+	bool empty();
+	void clear(); 
+	bool find(T & VAL);

```

Comment:
Why `T&` and not `const T&` ?

Reply:
It is being passed as a parameter inside the find function, can't do that if it was declared a const right?